### DISCUSSION TITLE
lazy loading when using leveldb as filer backend

### DISCUSSION THREAD

**Discussion by guol-fnst** [seaweedfs/seaweedfs#3945] [Ideas]
Avoid to loading too many leveldb files, in case of "too many open files"
1) Don't load any leveldb when volume process starts
2) When a file is accessed or uploaded or updated to this volume, load this leveldb index of  this volume
3) After a certain time , if this volume is not accessed , unload leveldb.

This feature may be a optional  configuration when starting volumes.

This feature may affect the performance for the first access of this volume, but it can reduce the opened files of leveldb. 
Expecially when there are too many small volumes.
Related https://github.com/seaweedfs/seaweedfs/discussions/3877

And here's another solution, don't load index files or leveldb files when index file's size is too small at the beginning. just load it when this volume is accessed. This solution still has less impact on performance.

**Comment by chrislusf** [comment]
Good idea! It would also save memory usage.

Let's keep the logic simple. Only off load based on last used time.

We can apply this logic only when the volume server is started in leveldb mode, which loading leveldb should be faster than in memory mode.

**Reply by guol-fnst** [reply]
Ok, I'll implement this feature these days.
Aso I need to confirm some details:
1) Add "-ldbTimeout int" configure to volume server. It's set to 0 defaultly, which means loading all the leveldb all the time. If it not equals to 0, then  leveldbs which is not accessed in ldbTimout days will be unloaded.
2) Should all leveldbs be loaded at first, and off load those which are timeout?  Or no leveldbs will be loaded, only loaded them when they are accessed.
  

Did I got your full opinions? If  I missunderstand, please poiont it out.

**Reply by chrislusf** [reply]
Sounds good.

**Reply by guol-fnst** [reply]
About piont 2, which is better?

**Reply by chrislusf** [reply]
2. if timeout is set, we can default to not loading at beginning.

**Reply by guol-fnst** [reply]
Got it.
But I find it's safer to unload leveldb right away after initial startup. Because some status need to be set into ldb at first.
New commit has been push into PR(https://github.com/seaweedfs/seaweedfs/pull/3958)

**Reply by guol-fnst** [reply]
And it will not cost too many fd and memory because volumes are not loaded in the same time.

`task_queue := make(chan os.DirEntry, 10*concurrency)`
